File: go/need-for-speed/need_for_speed.go
package speed

// TODO: define the 'Car' type struct

type Car struct {
    battery int
    batteryDrain int
    speed int
    distance int
}

// NewCar creates a new remote controlled car with full battery and given specifications.
func NewCar(speed, batteryDrain int) Car {
	newCar := Car{batteryDrain: batteryDrain, speed: speed, battery:100, distance: 0}
    return newCar
}

// TODO: define the 'Track' type struct

type Track struct {
    distance int
}

// NewTrack created a new track
func NewTrack(distance int) Track {
	newTrack := Track{distance: distance}
    return newTrack
}

// Drive drives the car one time. If there is not enough battery to drive one more time,
// the car will not move.
func Drive(car Car) Car {
    if car.battery >= car.batteryDrain {
		car.battery = car.battery - car.batteryDrain
    	car.distance = car.distance + car.speed
    }
    return car
}

// CanFinish checks if a car is able to finish a certain track.
func CanFinish(car Car, track Track) bool {
    numberOfDrives := car.battery/car.batteryDrain
	if track.distance <= car.speed*numberOfDrives {
        return true
    }
	return false
}

File: java/need-for-speed/src/main/java/NeedForSpeed.java
class NeedForSpeed {

    private int speed;
    private int batteryDrain;
    private int distanceDriven = 0;
    private int battery = 100;
    
    NeedForSpeed(int speed, int batteryDrain) {
        this.speed = speed;
        this.batteryDrain = batteryDrain;
    }

    public boolean batteryDrained() {
        return this.battery <= 0;
    }

    public int distanceDriven() {
        return this.distanceDriven;
    }

    public void drive() {
        if (this.battery >= this.batteryDrain) {
            this.distanceDriven += this.speed;
            this.battery -= this.batteryDrain;
        }
    }

    public static NeedForSpeed nitro() {
        return new NeedForSpeed(50, 4);
    }
}

class RaceTrack {

    private int distance;
    
    RaceTrack(int distance) {
        this.distance = distance;
    }

    public boolean tryFinishTrack(NeedForSpeed car) {
        while (!car.batteryDrained()) {
            car.drive();
            if (car.distanceDriven() >= this.distance) {
                return true;
            }
        }
        return false;
    }
}
